### PR TITLE
Lambda kernel example

### PR DESCRIPTION
**Pull request by erikzenker**
Shows how to use lambda function, std::function, and std::bind to construct an alpaka kernel.
It needs to be tested if this works for CUDA 8.0.

### CODE REVIEW

**Comment by BenjaminW3** [comment]
/home/travis/build/ComputationalRadiationPhysics/alpaka/example/helloWorldLambda/src/helloWorldLambda.cpp:136:71: error: 
      declaration shadows a local variable [-Werror,-Wshadow]
  ...ALPAKA_FN_ACC (Acc & acc, size_t const nExclamationMarks) -> void {
/home/travis/build/ComputationalRadiationPhysics/alpaka/example/helloWorldLambda/src/helloWorldLambda.cpp:131:18: note: 
      previous declaration is here
    const size_t nExclamationMarks = 10;

**Comment by BenjaminW3** [comment]
Aaaahhh... all GCC builds were green, even for CUDA 7.0. There was an error in the script executing CMake in the CI so the CUDA version was not set and the CUDA back-end was simply disabled. I am fixing this currently.

**Comment by erikzenker** [comment]
There is also missing the nvcc compiler flag `--expt-extended-lambda`. Where is the best place to put it ? Maybe [here](https://github.com/ComputationalRadiationPhysics/alpaka/blob/develop/alpakaConfig.cmake#L332) ?

**Comment by BenjaminW3** [comment]
Is this this flag required in CUDA 8.0? This flag should b conditionally be aded for CUDA 7.5+ then.

**Comment by erikzenker** [comment]
At least on my arch linux system I need to use this flag:
```
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Sun_Sep__4_22:14:01_CDT_2016
Cuda compilation tools, release 8.0, V8.0.44
```
error: __host__ or __device__ annotation on lambda requires --expt-extended-lambda nvcc flag
``

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+     * declaration might be useful when small kernels
+     * are written for testing or lambda functions
+     * allready exist.
+     */
+    const size_t nExclamationMarks = 10;
+    const size_t exclamationMarksMultiplier = 3;    
+    
+    auto const helloWorld(alpaka::exec::create<Acc>(
+        workdiv,
+	[&exclamationMarksMultiplier] ALPAKA_FN_ACC (Acc & acc, size_t const nExclamationMarks) -> void {
```
Comment (erikzenker) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: error: An extended __host__ __device__ lambda cannot capture variables by reference

Comment (BenjaminW3) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: I think that this is correct, as the reference to a host variable is not valid nor accessible on the device. You can only capture by copy (if it is allowed at all).

Comment (psychocoderHPC) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: Yes only copy by value or a this pointer is allowed. As I know in the case of a this pointer the full class is copied to the device.

**Comment by erikzenker** [comment]
I addressed the code review issues.

File: alpakaConfig.cmake [hunk 2]
```diff
@@ -329,11 +329,12 @@ IF(ALPAKA_ACC_GPU_CUDA_ENABLE)
 
                 SET(CUDA_PROPAGATE_HOST_FLAGS ON)
 
+
                 IF(ALPAKA_CUDA_VERSION VERSION_EQUAL 8.0)
                     LIST(APPEND CUDA_NVCC_FLAGS "-Wno-deprecated-gpu-targets")
                 ENDIF()
 
-                LIST(APPEND CUDA_NVCC_FLAGS "-arch=${ALPAKA_CUDA_ARCH}")
+                LIST(APPEND CUDA_NVCC_FLAGS "-arch=${ALPAKA_CUDA_ARCH} --expt-extended-lambda")
```
Comment (BenjaminW3) [inline alpakaConfig.cmake]: Please add this flag in an extra line and only 'IF(NOT ALPAKA_CUDA_VERSION VERSION_LESS 7.5)'

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	    auto linearizedGlobalThreadIdx = alpaka::idx::mapIdx<1u>(globalThreadIdx,
+								      globalThreadExtent);
+                                                          
+	    printf("[z:%u, y:%u, x:%u][linear:%u] Hello world from a lambda",
+		   static_cast<unsigned>(globalThreadIdx[0]),
+		   static_cast<unsigned>(globalThreadIdx[1]),
+		   static_cast<unsigned>(globalThreadIdx[2]),
+		   static_cast<unsigned>(linearizedGlobalThreadIdx[0]));
+
+	    for(size_t i = 0; i < nExclamationMarks * exclamationMarksMultiplier; ++i){
```
Comment (BenjaminW3) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: This should be nExclamationMarksAsArg

File: .travis.yml [hunk 4]
```diff
@@ -728,6 +728,7 @@ script:
     # Build and execute all examples.
     - if [ "${ALPAKA_CI_ANALYSIS}" == "OFF" -a "${ALPAKA_ACC_GPU_CUDA_ONLY_MODE}" == "OFF" ] ;then ./travis/compileExec.sh "example/bufferCopy" ./bufferCopy ;fi
     - if [ "${ALPAKA_CI_ANALYSIS}" == "OFF" -a "${ALPAKA_ACC_GPU_CUDA_ONLY_MODE}" == "OFF" ] ;then ./travis/compileExec.sh "example/helloWorld" ./helloWorld ;fi
+    - if [ "${ALPAKA_CI_ANALYSIS}" == "OFF" -a "${ALPAKA_ACC_GPU_CUDA_ONLY_MODE}" == "OFF" -a ${ALPAKA_CUDA_VER_MAJOR} > 7 -a ${ALPAKA_CUDA_VER_MINOR} > 0] ;then ./travis/compileExec.sh "example/helloWorldLambda" ./helloWorldLambda ;fi
```
Comment (BenjaminW3) [inline .travis.yml]: This is not correct because it excludes cuda 8.0. (maj>7) || ((maj == 7) && (min >5)) should be better.

Comment (BenjaminW3) [inline .travis.yml]: I am no bash expert but the syntax is not supported: The builds say `/home/travis/build.sh: line 45: [: missing `]'` but did not get red (I do not know why).
You may need to use double parenthesis like I did in other places to make the arithmetic comparison work.

Comment (erikzenker) [inline .travis.yml]: There is a space missing in front of `]` ...I try to fix this in the next days.

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+     * to wrap allready existing code into a
+     * std::function and provide it to the alpaka 
+     * library.
+     */
+    auto const hiWorld (alpaka::exec::create<Acc> (
+        workdiv,
+        std::function<void(Acc&, size_t)>( hiWorldFunction<Acc> ),
+        nExclamationMarks));
+
+
```
Comment (BenjaminW3) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: please remove this additional newline

**Comment by BenjaminW3** [comment]
Looks good! Please squash all the commits when everything is green.

**Pull request merged**